### PR TITLE
Preform RBAC checks before connecting to registered OpenSSH nodes

### PR DESCRIPTION
**Pull request by capnspacehook**
Part of implementing https://github.com/gravitational/teleport/pull/19261. 

Preform RBAC checks before connecting to registered OpenSSH nodes. The RBAC logic assumes that SSH connections to registered OpenSSH nodes will always be proxied and terminated, which will come in a future PR.

### CODE REVIEW

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go:407]: That's a good idea, I'll do that

Comment (capnspacehook) [reply lib/srv/authhandlers.go]: Done

Comment (capnspacehook) [reply lib/srv/authhandlers.go]: Oops yes, meant to do that

Comment (capnspacehook) [reply lib/srv/authhandlers.go]: Done

File: lib/srv/authhandlers.go
Comment (jakule) [inline lib/srv/authhandlers.go:551]: Can we change the abstraction here? This looks hacky in my opinion. You call `canLoginWithoutRBAC` but when `AuthorizeWithoutRBAC` is set RBAC logic won't be called at all only for testing purposes.
I think it would be cleaner to extract `canLoginWithoutRBAC` and `canLoginWithRBAC` as interfaces and provide a different implementation in the production code and tests. In case you forgot to set the code will panic which I think it's better then calling mock by mistake.

Before we go that path, what stops us from running RBAC check in tests? Sounds like something without many side effects.

File: lib/srv/authhandlers.go
Comment (jakule) [inline lib/srv/authhandlers.go:99]: Check my comment below.

File: lib/srv/forward/sshserver.go
Comment (jakule) [inline lib/srv/forward/sshserver.go:167]: Comment please

File: lib/srv/forward/sshserver.go
Comment (jakule) [inline lib/srv/forward/sshserver.go:227]: Comment, please.

File: lib/srv/authhandlers.go
Comment (jakule) [inline lib/srv/authhandlers.go:426]: Shouldn't we return an error when `h.c.TargetNode == nil && h.c.TargetNode.GetSubKind() == types.KindOpenSSHNode` instead of allowing access without RBAC?

Also, how this case differs from the default case `h.c.Component != teleport.ComponentForwardingNode`?

**Review by jakule — CHANGES_REQUESTED** [review]

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go:426]: The first case is impossible if `.c.TargetNode == nil`, did you mean `.c.TargetNode != nil && h.c.TargetNode.GetSubKind() != types.KindOpenSSHNode`?

If `h.c.TargetNode` is `nil`, the node is not registered so we can't preform a RBAC check.
If the `AuthHandlers` are running for a regular SSH node, a RBAC check should always be preformed.

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go:551]: We can run an RBAC check in the tests I added, but I don't think it would be necessary. There are already other tests that confirm that RBAC works correctly, so my thought was instead of needlessly complicating the test by creating users, roles, etc so RBAC will work, just test the a RBAC check happens or doesn't happen.

File: lib/srv/authhandlers.go
Comment (jakule) [reply lib/srv/authhandlers.go:426]: NVM

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go]: Done

File: lib/srv/forward/sshserver.go
Comment (capnspacehook) [reply lib/srv/forward/sshserver.go:168]: Done

File: lib/srv/forward/sshserver.go
Comment (capnspacehook) [reply lib/srv/forward/sshserver.go:229]: Done

File: lib/srv/authhandlers_test.go
Comment (jakule) [inline lib/srv/authhandlers_test.go:130]: Do we need two booleans for that? Maybe
```
shouldAuthWithRBAC: [true/false]
```
is enough?

File: lib/srv/authhandlers.go
Comment (jakule) [inline lib/srv/authhandlers.go:99]: Since we're embedding the interface for RBAC, why do we still need these function pointers here?

File: lib/srv/authhandlers.go
Comment (jakule) [inline lib/srv/authhandlers.go:570]: Do we need `AuthHndlers` struct to just use a logger and one function? Can we reduce the requirements here?

File: lib/srv/authhandlers.go
Comment (jakule) [inline lib/srv/authhandlers.go:557]: This interface is extracted to enable testing, but it's not created only for testing. 
```suggestion
// a node.
```

File: lib/srv/authhandlers.go
Comment (jakule) [inline lib/srv/authhandlers.go:579]: As we talked about earlier, please remove this redirection. It's not needed anymore since the `AuthHandler` can accept different implementations of `loginChecker` interface.

**Review by jakule — CHANGES_REQUESTED** [review]

File: lib/srv/authhandlers_test.go
Comment (capnspacehook) [reply lib/srv/authhandlers_test.go:130]: I want to ensure either `canLoginWithoutRBAC` or `canLoginWithRBAC` is called, and I need two booleans to do that.

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go]: Oops, removed

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go]: Updated so only logger, config and small interface is needed, does it look good now?

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go]: Done

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go]: Done

**Comment by capnspacehook** [comment]
@rosstimothy can I have another review please?

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go:572]: I guess that's true. I also thought `canLoginWithoutRBAC` was called multiple times, but it's only called once. So removing it shouldn't be a big deal.

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go]: Done

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go]: Done

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go]: Done

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go]: Done

File: lib/srv/authhandlers_test.go
Comment (capnspacehook) [reply lib/srv/authhandlers_test.go]: Done

File: lib/srv/authhandlers_test.go
Comment (capnspacehook) [reply lib/srv/authhandlers_test.go]: Done

File: lib/srv/forward/sshserver.go
Comment (capnspacehook) [reply lib/srv/forward/sshserver.go]: Done

File: lib/srv/forward/sshserver.go
Comment (capnspacehook) [reply lib/srv/forward/sshserver.go]: Done

File: lib/srv/authhandlers_test.go
Comment (capnspacehook) [reply lib/srv/authhandlers_test.go]: Done

File: lib/srv/authhandlers.go
Comment (jakule) [inline lib/srv/authhandlers.go:420]: @capnspacehook Can you please add a comment that this check is needed even if we don't need the full RBAC to check to make it less likely that someone will remove it in the future?

**Review by jakule — APPROVED** [review]
LGTM with a few nits

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go:420]: Good call, done

File: lib/srv/authhandlers.go
Comment (capnspacehook) [reply lib/srv/authhandlers.go:422]: I did that to keep logging behavior the same as before. Calling `authorityForCert` is essentially all that `canLoginWithoutRBAC` did, and if it failed the error would be logged in the same way below.

File: lib/srv/authhandlers_test.go
Comment (capnspacehook) [reply lib/srv/authhandlers_test.go:93]: Good catch, done

**Review by rosstimothy — COMMENTED** [review]
LGTM thanks for addressing all of my comments. I'm going to hold off on approving until the RFD has been approved and merged though.

**Comment by capnspacehook** [comment]
@rosstimothy RFD is in the merge queue currently, so this should be good to merge now

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**
